content/blog/2022-02-22-profiling.md

```diff
@@ -0,0 +1,208 @@
+
+Figure (a) has two back edges I->A and inner E->B. We add dummy edges A->B and E->I, and remove the back edge E->B. Note that A is the entry block, and I is the exit block in this CFG. So the dummy edges we add are the edge from the entry block to the destination node of the back edge, and the edge from the source node of the back edge to the exit block. Through this transformation, general CFGs are downgraded to the previous CFG problem we already solved. 
+
+<p align="center">
+<img src="https://res.cloudinary.com/dxzx2bxch/image/upload/v1646243822/cs6120/5_ajv4lx.png" alt="alt_text" title="image_tooltip" style="zoom:50%;" />
+</p>
+<p align="center">
+<strong>Removing back edge and creating dummy edges for general CFGs</strong>
+</p>
+
+[Discussion] 
+An intuitive way to understand why adding dummy edges works: the original back edge E->B is denoted as E->I->A->B in the transformed graph with the help of dummy edges. 
+
+
+
+### Results
+
+![alt_text](https://res.cloudinary.com/dxzx2bxch/image/upload/v1646243822/cs6120/6_xpyzbz.png "image_tooltip")
+
+
+Table 1 shows the experiment results on runtime overhead of path profiling (PP) and edge profiling (QPT). The runtime overhead of path profiling is 30.9%, while the overhead of edge profiling is 16.1%. 
+
+
+#### Is the overhead acceptable?
+
+We discussed it with the conclusion that it depends (oh, the ultimate answer!). 
+
+It depends on the application and how frequently you need to use path profiling. For example, if we use it in JIT compilation and only profile it once, then 30% overhead is acceptable for its improving performance. 
+
+It’s also worth noting that the result presented in this paper might be the worst-case scenario of runtime overhead because the compiled SPEC benchmarks execute very fast. But in real cases, we may profile at the bytecode level. Then the overhead ratio might be lower because bytecode execution is slower. 
+
+
+### Limitations
+
+
+#### Is it efficient enough?
+
+We need to run edge profiling first and determine where we will put our instrumentation. Only after that, can we run path profiling. This additional overhead makes readers think that the Ball-Larus method would not be as efficient as it claims in the paper. 
+
+Some other works also raise questions on it: Kumar et al. [1] claim that methods like Ball-Larus algorithm take significant time to identify profile points in the program, cannot be used in dynamic (JIT) compilation. 
+
+We think the phase of identifying instrumentation could be accelerated by only analyzing a tiny and representative sample of the entire program because we only need a rough idea for the edge profiling result. There is also an interesting idea to adopt static program analysis methods like “constant propagation” to know some deterministic value range of certain variables to know additional information about edge profiling without any runtime analysis. 
```

Comment:
It's a good point that the initial edge profile is really just "nice to have" rather than critical for the path profiling technique's correctness/success. So it would be fine to have a "quick and dirty" approximate edge profile; the only thing that will suffer is the efficiency (not the correctness) of the eventual path profile.